src/Access/IAccessStorage.cpp

```diff
@@ -514,6 +514,14 @@ bool IAccessStorage::areCredentialsValid(
     if (credentials.getUserName() != user.getName())
         return false;
 
+    if (user.valid_until)
+    {
+        const time_t now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+
+        if (now > user.valid_until)
+            return false;
```

Comment:
This will throw `ErrorCodes::WRONG_PASSWORD`, which is probably not the correct error

Reply:
Also, shouldn't we remove expired users after some time?

Reply:
The `VALID UNTIL` clause is a way to set the expiration date for the credentials of the user, similar to `VALID UNTIL` in PostgreSQL. So after the expiration date has passed not the user but the password becomes invalid, so the `ErrorCodes::WRONG_PASSWORD` error sounds valid and we also don't need to necessarily remove those users.